src/sensor_ads1220.c

```diff
@@ -0,0 +1,206 @@
+#include "spicmds.h" // spidev_transfer
+
+struct ads1220_adc {
+    struct timer timer;
+    uint32_t rest_ticks;
+    struct gpio_in data_ready;
+    struct spidev_s *spi;
+    uint8_t flags, data_count;
+    struct sensor_bulk sb;
+};
+
+// Flag types
+enum {
+    FLAG_PENDING = 1 << 0
+};
+
+#define BYTES_PER_SAMPLE 4
+
+static struct task_wake wake_ads1220;
+
+/****************************************************************
+ * ADS1220 Sensor Support
+ ****************************************************************/
+#define MAX_SPI_READ_TIME timer_from_us(150)
+
+static inline uint8_t
+is_flag_set(const uint8_t mask, struct ads1220_adc *ads1220)
+{
+    return !!(mask & ads1220->flags);
+}
+
+static inline void
+set_flag(uint8_t mask, struct ads1220_adc *ads1220)
+{
+    ads1220->flags |= mask;
+}
+
+static inline void
+clear_flag(uint8_t mask, struct ads1220_adc *ads1220)
+{
+    ads1220->flags &= ~mask;
+}
+
+// Event handler that wakes wake_ads1220() periodically
+static uint_fast8_t
+ads1220_event(struct timer *timer)
+{
+    struct ads1220_adc *ads1220 = container_of(timer, struct ads1220_adc,
+                                                timer);
+    set_flag(FLAG_PENDING, ads1220);
+    sched_wake_task(&wake_ads1220);
+    return SF_DONE;
+}
+
+// Helper code to reschedule the ads1220_event() timer
+static void
+ads1220_reschedule_timer(struct ads1220_adc *ads1220)
+{
+    irq_disable();
+    set_flag(FLAG_PENDING, ads1220);
+    ads1220->timer.waketime = timer_read_time() + ads1220->rest_ticks;
+    sched_add_timer(&ads1220->timer);
+    irq_enable();
+}
+
+int8_t
+ads1220_is_data_ready(struct ads1220_adc *ads1220) {
+    return gpio_in_read(ads1220->data_ready) == 0;
+}
+
+// Add a measurement to the buffer
+static void
+add_sample(struct ads1220_adc *ads1220, uint_fast32_t counts)
+{
+    ads1220->sb.data[ads1220->sb.data_count] = counts;
+    ads1220->sb.data[ads1220->sb.data_count + 1] = counts >> 8;
+    ads1220->sb.data[ads1220->sb.data_count + 2] = counts >> 16;
+    ads1220->sb.data[ads1220->sb.data_count + 3] = counts >> 24;
+    ads1220->sb.data_count += BYTES_PER_SAMPLE;
+}
+
+static void
+flush_samples(struct ads1220_adc *ads1220, uint8_t oid)
+{
+    if ((ads1220->sb.data_count + BYTES_PER_SAMPLE) >
+            ARRAY_SIZE(ads1220->sb.data)) {
+        sensor_bulk_report(&ads1220->sb, oid);
+    }
+}
+
+// ADS1220 ADC query
+void
+ads1220_read_adc(struct ads1220_adc *ads1220, uint8_t oid)
+{
+    if (!ads1220_is_data_ready(ads1220)) {
+        ads1220_reschedule_timer(ads1220);
+        return;
+    }
+
+    // data is ready
+    uint8_t msg[3] = {0, 0, 0};
+    uint32_t start_time = timer_read_time();
+    spidev_transfer(ads1220->spi, 1, sizeof(msg), msg);
+    uint32_t time_diff = timer_read_time() - start_time;
+
+    if (time_diff >= MAX_SPI_READ_TIME) {
+        // some IRQ delayed this read so much that its unusable
+        shutdown("ADS1220 read timing error, read took too long");
```

Comment:
I don't think we should issue a `shutdown()` in any of this sensor code.  The code should encode an error of some kind in the sensor reading and then let the python code figure out how to handle the event.

Same comment for the sensor_hx71x.c code.

Reply:
I agree with you in spirit. I don't like the idea of this ruining a print. I had an implementation of this but it was complex.

If something goes wrong with these chips (particularly the HX71x [reference](https://github.com/prusa3d/Prusa-Firmware-Buddy/blob/7c651de64ad3973f8faa10d769d3156d9b2e288a/src/common/hx717.cpp)) the recovery path is to:
* Stop measurements
* Put the chip into a reset state for some minimum time
* Bring the chip out of reset and re-configure it
  * For the HX71x this require running the measurement loop and discarding the first sample
* Restart measurements

While this reset is going on, any process that's consuming measurements now doesn't get them. This might be an issue for the consumer. E.g. calibrating the scale: getting an error should abort the gcode command. So I cant really hide the reset event from downstream systems. Either its communicated as a value, or by disconnecting any clients of the read process.

This also raises the specter of reset loops. When a chip has ESD issues it will fault frequently. 1 or more times per second. How many faults before we shutdown the system?

(Resets on the MCU during probing are handled by the watch dog thread in the load cell endstop. So that part at least is resilient already.)

Reply:
Just to be clear, the concern is not with raising an error - the concern is in using `shutdown()` to indicate the error.  A `shutdown()` from the mcu provides little flexibility in handling and reporting the situation - heaters, motors, displays, leds, etc. are all disabled and it is often necessary to analyse the log to determine what happened.

We really want to propagate the error back to the host code so that it can handle it.  If the host code raises an error (or shutdown) that's fine.

You can take a look at how the adxl345 code checks for error conditions and introduces error codes into the sample stream.  The adxl345.py code then checks for those error codes and propagates that information via statistics.